### DISCUSSION TITLE
Welcome to bitcoin-faucet-shell Discussions!

### DISCUSSION THREAD

**Discussion by jsarenik** [jsarenik/bitcoin-faucet-shell#6] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by xniper12** [comment]
It says slow down 
tb1qdwxa65ffyq677lf8zec3eh6lwz9jak3gq0j7au

**Reply by cryptonian12** [reply]
Same here, 
Said slow down.

**Reply by carnhofdaki** [reply]
Try https://delvingbitcoin.org/t/proof-of-work-based-signet-faucet/937 which works much better.

**Reply by jsarenik** [reply]
> It says slow down tb1qdwxa65ffyq677lf8zec3eh6lwz9jak3gq0j7au

Try now. And see https://github.com/jsarenik/bitcoin-faucet-shell/discussions/10

**Comment by lilin998** [comment]
502

**Reply by carnhofdaki** [reply]
Let's discuss 502 and other numbers, what do you think?
Come to the kitchen: https://kitchen.anyone.eu.org/

PS: https://kitchen.anyone.eu.org/tnvctc/new/mp3/2-07_A_language_of_life.mp3

**Comment by jsarenik** [comment]
Some real-life [benchmark](https://mempool.space/signet/tx/2f2187263e2bbfb4af116ddcf6df4d0558af0e3b7df945ec6a6abb79a9382b5f) that happened when I was playing with refreshing the wallets and regaining a few megabytes of the diskspace.

See https://github.com/bitcoin/bitcoin/issues/31816

**Comment by jsarenik** [comment]
The previous recently added comment was spam (and promotes scam). The user has been blocked from this discussion and reported to GitHub. The entry is hidden though still available for observation.

> Don't waste your time on lusers. Sometimes you're ahead, sometimes you're behind. The race is long and, in the end, it's only with yourself.
>
> https://www.bublina.eu.org/x31/netvorba/sfx/
>
> https://soundcloud.com/jasanitto/on-the-firewalls

**Reply by jsarenik** [reply]
Thank you for contacting GitHub Support. We wanted to let you know that we've received your message. We are experiencing high volumes and therefore, you may experience longer than normal wait times. In the meantime, you might find the answer to your question or a self-service solution to your problem in the GitHub Docs. Feel free to check it out!

You can also post a question for other GitHub users in the GitHub Community. The Community can't help with account issues, like reinstatement requests, or with getting your ticket looked at faster.


Ticket ID: 3244245



If you would like to add any additional information to your ticket, feel free to reply to this email.

Thank you!
The GitHub Support Team

This email is a service from GitHub Support.
[NYL0M7-9ZW47]

**Reply by jsarenik** [reply]
Everything sorted out. Professionally. Thanks GitHub!I wrote them this in the beginning;

> The content of reported comment is scam and spam.
>
> On the other hand this faucet is meant to spread knowledge about internal details of Bitcoin using its testing network called Signet where anyone can get coins for free and learn to understand how Bitcoin works under the hood.

**Comment by jsarenik** [comment]
Currently the only limits are following:

 - Cloudflare Turnstile
 - after 2016 outputs in payout transaction (never really happens, but let's try!) no more requests are accepted from the same IP address (`Please slow down` error shows)

**Reply by jsarenik** [reply]
And it works:

![Screenshot_20250307-063653.png](https://github.com/user-attachments/assets/8605dc4c-3c6a-4279-9dcd-5871150b0ad0)

https://mempool.space/signet/tx/67714744d373874f4a14644addeb7e5ee3c38d7be2187e77c465f311622d065c?mode=details

**Reply by jsarenik** [reply]
1216 requests served in https://mempool.space/signet/tx/520a001d431e6f27dc79a971129bcd92eb8e1c780a75bd661dff3117870909d6

**Reply by jsarenik** [reply]
2053 requests served in https://mempool.space/signet/tx/aeb582b2a142d2ef45bfda383b3c51c4cb13ab6172e7bd689de2721e52316538?mode=details

**Reply by lushuguang1223** [reply]
Hello, jsarenik
Since one day last week, there has been a sudden increase in faucet requests, thousands of them, are you testing?
![requests](https://github.com/user-attachments/assets/1356dfc2-87bd-4ebe-b0b7-326622c4da11)

**Reply by jsarenik** [reply]
@lushuguang1223 wrote:
> Since one day last week, there has been a sudden increase in faucet requests, thousands of them, are you testing?

I would say it may be related to [switching back to V2 transactions](https://github.com/jsarenik/bitcoin-faucet-shell/discussions/17)

**Reply by jsarenik** [reply]
@lushuguang1223 see also https://github.com/jsarenik/bitcoin-faucet-shell/issues/4#issuecomment-2708724663

**Reply by jsarenik** [reply]
@lushuguang1223 I am not doing any testing now. The requests are all from the outside and each had to pass Cloudflare Turnstile. ~Except for the example requests which are located early (right after OP_RETURN, i.e. vout 1) in every single payout transaction.~

**Comment by jsarenik** [comment]
The latest thing is that number of always-present early outputs was extendedvso that the number of faucet requests served in this payout matches the number shown by mempool space (when bigger than 210).

![Screenshot_20250313-172344.png](https://github.com/user-attachments/assets/4e8e6202-f33d-45e6-8962-a11fabdddaaa)

**Comment by jsarenik** [comment]
This faucet is my keep-alive monitor. The internet connection I use is the cheapest available and it would be paid for anyway. Thanks to `sch_cake` Linux module all users of the network are happy. All the used hardware is recycled and works well since 2012. May it work like it is another 10 years! :)

**Comment by jsarenik** [comment]
`maxmempool=10`

**Comment by jsarenik** [comment]
API was reactivated in https://github.com/jsarenik/bitcoin-faucet-shell/commit/f6babb1f89e320ca558c2aada35809cb0c517da6

**Reply by jsarenik** [reply]
https://mempool.space/signet/tx/c45ede7f002dd3d106d1f4427a3385dcb506f40957071c3780d909ea83448375?mode=details

**Reply by jsarenik** [reply]
Even though the API without further checks works now, the humans using the turnstile are prioritized. As long as there is space in the queue since the turnstile-checking bots do exist as well. Let's wait until it settles down now (from over 19,000 queued requests).

**Reply by jsarenik** [reply]
https://mempool.space/signet/tx/06b05343f040e6214560a1f4331984ec852a0c00ac83993ad0123ebdd4c5bf0f?mode=details

**Reply by jsarenik** [reply]
It was turned off again in https://github.com/jsarenik/bitcoin-faucet-shell/commit/a5cd1f2eb89b7237ff21cbd904860b0128e156ad

**Comment by jsarenik** [comment]
An additional error condition was added that makes the faucet deny any new output if the current queue size is bigger than `sfsm` (now 4032). "loli"

Uncommitted in git repo yet.

**Comment by jsarenik** [comment]
Feel free to reimplement in your favorite language and run a faucet yourself. wink wink :)

**Comment by jsarenik** [comment]
Timeline (as proposed in Bitcoin Grant Application, today approved by [OpenSats](https://opensats.org/))

1. Jul 2025 - start documenting the pillars this faucet stands on right now, make notes but
do not change any code yet because it works as is
2. Aug 2025 - go through the notes and propose improvements via GitHub issues
3. Sep 2025 - start doin PRs to address the issues, hopefully widening the user community
slowly
4. Oct 2025 - start the faucet on a leftover Chaincode FOSS 2024 custom signet network
which is alive and mining 5 blocks per hour (minimal difficulty) on my laptop
5. Nov 2025 - clean and fast code of the bitcoin-shell-faucet that is easily deployable by
anyone anywhere
6. Dec 2025 - everyone is using the faucet to learn about Bitcoin

**Reply by jsarenik** [reply]
[Alt Signet Faucet](https://signetfaucet.bublina.eu.org/) adds requested addresses to a tmpfs file-based queue and replaces (RBF) the transaction with more outputs every 30 seconds as the requests come.

Only interactive requests are possible (see now defunct `contrib/getcoins.sh` for API usage example).

The replaced transaction is (usually; sometimes race-conditioned) 25th in a mempool chain, disallowing further movement until in a block (makes replacements easier, if there was no 25-chain a child could spend itself in an all-fee tx rendering future replacements very test-coins-expensive)

The wallet itself is WIF-only (no Xprvs) and is being refreshed every hour. The wallet's refreshed sqlite3 DB file resides in tmpfs (Linux) so it is super-fast and does not grow in time.

Sometimes there are still very detail race-conditions in my scripts but in long-term it already works and lives for more than four months.

Rate-limiting (one of latest changes) is done by atomic mkdir so there are no concurrent requests for the same IP at the same time. rmdir(1) on tmpfs clears the lock when requested address is written into its own tiny plain-text file, again on tmpfs (in RAM).

The requested addresses are decoded asynchronously into ScriptPubKey (hex) by Steven Roose's hal(1). P2PK works both for uncompressed and compressed hexadecimal public keys (see `contrib/sff/signetfaucet.sh` in the git repo).

The changes needed for deployment on other test networks would be minimal and you could easily use the scripts to run another faucet for Signet as they are now. I will be happy to talk and answer any questions about it.

The faucet runs on a recycled old fanless laptop which is the HTTPS origin (and serves other Bitcoin-related sites, so the faucet is not the only, but important service), but gets proxied by Cloudflare in order to hide the real public IP address and slightly lower the number of attacks. Could run on anything like Raspberry Pi under Linux.

HTML is static and shows up as long as the server is online. My main home router uses the Linux's [cake scheduler](https://www.bufferbloat.net/projects/codel/wiki/Cake/) which ensures that apart from rare ISP or hardly-seen electricity outages everything works well and all the client users of the 50/10 Mbps network are happy.

**Comment by jsarenik** [comment]
Hash-client-side-only address setting now available. Try https://alt.signetfaucet.com/#tb1qg3lau83hm9e9tdvzr5k7aqtw3uv0dwkfct4xdn
